primitives/src/block.rs

```diff
@@ -15,6 +16,78 @@ use hashes::{sha256d, HashEngine as _};
 
 use crate::merkle_tree::TxMerkleNode;
 use crate::pow::CompactTarget;
+use crate::transaction::Transaction;
+
+/// Bitcoin block.
+///
+/// A collection of transactions with an attached proof of work.
+///
+/// See [Bitcoin Wiki: Block][wiki-block] for more information.
+///
+/// [wiki-block]: https://en.bitcoin.it/wiki/Block
+///
+/// ### Bitcoin Core References
+///
+/// * [CBlock definition](https://github.com/bitcoin/bitcoin/blob/345457b542b6a980ccfbc868af0970a6f91d1b82/src/primitives/block.h#L62)
+#[cfg(feature = "alloc")]
+#[derive(PartialEq, Eq, Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Block<V = Unchecked>
+where
+    V: Validation,
+{
+    /// The block header
+    pub header: Header,
+    /// List of transactions contained in the block
+    pub txdata: Vec<Transaction>,
+    /// Validation marker.
+    marker: PhantomData<V>,
+}
+
+// TODO: Move this to `primitives`.
+#[cfg(feature = "alloc")]
+impl Block<Unchecked> {
+    /// Constructs a new `Block` without doing any validation.
+    pub fn new_unchecked(header: Header, txdata: Vec<Transaction>) -> Block<Unchecked> {
+        Block { header, txdata, marker: PhantomData::<Unchecked> }
+    }
+
+    /// Ignores block validation logic and just assumes you know what you are doing.
+    ///
+    /// You should only use this function if you trust the block i.e., it comes from a trusted node.
+    pub fn assume_checked(self) -> Block<Checked> {
```

Comment:
Perhaps a `#[must_use]` maybe?

Reply:
We could put one, but since `Block` isn't `Copy` and `self` is taken by value, users will quickly realize that they need to keep the return value.